pyiceberg/table/__init__.py

```diff
@@ -3253,6 +3253,15 @@ def __init__(self, tbl: Table) -> None:
         except ModuleNotFoundError as e:
             raise ModuleNotFoundError("For metadata operations PyArrow needs to be installed") from e
 
+    def _snapshot(self, snapshot_id: Optional[int] = None) -> Optional[Snapshot]:
+        if snapshot_id:
+            if snapshot := self.tbl.metadata.snapshot_by_id(snapshot_id):
+                return snapshot
+            else:
+                raise ValueError(f"Cannot find snapshot with ID {snapshot_id}")
```

Comment:
nit: if the return value is `Optional[Snapshot]`, maybe the function should not raise `ValueError`

Reply:
Hi @kevinjqliu thanks for the review. 

I thought about this, and I stand by this behavior / type annotation. This is my rationale:

1. If the user passes in `snapshot_id` - that's an indication that the user wants to look up a specific snapshot_id (instead of using the current one). Then we should look up snapshot_by_id and see if we can find the corresponding Snapshot and if we can't find the Snapshot, we should raise.
2. The output arg type is still `Optional[Snapshot]` even if we raise in the above behavior, because tbl.metadata.current_snapshot() returns `Optional[Snapshot]`. https://github.com/apache/iceberg-python/blob/5039b5d70644bc06c98349090912c6e9066d3ea1/pyiceberg/table/metadata.py#L295

Reply:
And I believe the above reason is because a newly created table isn't required to have a snapshot

Reply:
👍 make sense, thanks for the explanation

Reply:
No problem, and thank you again for the review!

Reply:
I agree that we should raise an error when the snapshot cannot be found. What do you tink of updating the signature to `Snapshot`, and also raise an exception when there is no current snapshot?

Reply:
@Fokko Sure - I thought it would be more correct to return an empty metadata table (entries, partitions, etc) if there's no snapshot in the table than raising an Exception, but this way I think we avoid extra `if` statements in each of the metadata table generating methods.